In [1]:
import pandas as pd

base_path = '/content/drive/MyDrive/healthcarefraud_detection/data/raw'

train = pd.read_csv(f'{base_path}/Train.csv')
test = pd.read_csv(f'{base_path}/Test.csv')
train_bene = pd.read_csv(f'{base_path}/Train_Beneficiarydata.csv')
test_bene = pd.read_csv(f'{base_path}/Test_Beneficiarydata.csv')
train_inp = pd.read_csv(f'{base_path}/Train_Inpatientdata.csv')
test_inp = pd.read_csv(f'{base_path}/Test_Inpatientdata.csv')
train_out = pd.read_csv(f'{base_path}/Train_Outpatientdata.csv')
test_out = pd.read_csv(f'{base_path}/Test_Outpatientdata.csv')



In [2]:
from IPython.display import display
import pandas as pd

def column_sample_report_transposed(df, df_name, n=5):
    """
    Generates a transposed tabular report of columns, their types, and sample values.

    Args:
        df (pd.DataFrame): The DataFrame to report on.
        df_name (str): The name of the DataFrame for the report title.
        n (int): The number of unique values to sample per column.
    """
    print(f"\n=== {df_name} ===")

    report_data = {"Information Type": ["Column Name", "Data Type"] + [f"Sample {i+1}" for i in range(n)]}

    for col in df.columns:
        dtype = df[col].dtype
        vals = df[col].dropna().unique()
        sample = list(vals[:n]) + [None] * (n - len(vals[:n])) # Pad with None if less than n samples

        report_data[col] = [col, dtype] + sample

    report_df = pd.DataFrame(report_data)
    report_df = report_df.set_index("Information Type")
    display(report_df)

# Repeat for each DataFrame you care about using the new function
column_sample_report_transposed(train_inp, "Train_Inpatientdata")
column_sample_report_transposed(train_out, "Train_Outpatientdata")
column_sample_report_transposed(train_bene, "Train_Beneficiarydata")
column_sample_report_transposed(train, "Train")


=== Train_Inpatientdata ===


BeneID   ClaimID  ClaimStartDt  ClaimEndDt  Provider  \
Information Type                                                            
Column Name          BeneID   ClaimID  ClaimStartDt  ClaimEndDt  Provider   
Data Type            object    object        object      object    object   
Sample 1          BENE11001  CLM46614    2009-04-12  2009-04-18  PRV55912   
Sample 2          BENE11011  CLM66048    2009-08-31  2009-09-02  PRV55907   
Sample 3          BENE11014  CLM68358    2009-09-17  2009-09-20  PRV56046   
Sample 4          BENE11017  CLM38412    2009-02-14  2009-02-22  PRV52405   
Sample 5          BENE11018  CLM63689    2009-08-13  2009-08-30  PRV56614   

                  InscClaimAmtReimbursed  AttendingPhysician  \
Information Type                                               
Column Name       InscClaimAmtReimbursed  AttendingPhysician   
Data Type                          int64              object   
Sample 1                           26000           PHY390922   
Sample 2                            5000           PHY318495   
Sample 3                           10000           PHY372395   
Sample 4                            8000           PHY369659   
Sample 5                            6000           PHY379376   

                  OperatingPhysician  OtherPhysician  AdmissionDt  ...  \
Information Type                                                   ...   
Column Name       OperatingPhysician  OtherPhysician  AdmissionDt  ...   
Data Type                     object          object       object  ...   
Sample 1                   PHY318495       PHY324689   2009-04-12  ...   
Sample 2                   PHY392961       PHY349768   2009-08-31  ...   
Sample 3                   PHY398258       PHY402711   2009-09-17  ...   
Sample 4                   PHY402711       PHY423511   2009-02-14  ...   
Sample 5                   PHY347494       PHY357982   2009-08-13  ...   

                  ClmDiagnosisCode_7  ClmDiagnosisCode_8  ClmDiagnosisCode_9  \
Information Type                                                               
Column Name       ClmDiagnosisCode_7  ClmDiagnosisCode_8  ClmDiagnosisCode_9   
Data Type                     object              object              object   
Sample 1                        2724               19889                5849   
Sample 2                       25062               40390                4019   
Sample 3                        5119               29620               20300   
Sample 4                        2948               25000               25002   
Sample 5                        2851                2809               56210   

                  ClmDiagnosisCode_10  ClmProcedureCode_1  ClmProcedureCode_2  \
Information Type                                                                
Column Name       ClmDiagnosisCode_10  ClmProcedureCode_1  ClmProcedureCode_2   
Data Type                      object             float64             float64   
Sample 1                        42732              7092.0              4139.0   
Sample 2                        30981               331.0              4019.0   
Sample 3                        32723              3893.0              2724.0   
Sample 4                        V4364               863.0              7904.0   
Sample 5                         V153              4576.0              5849.0   

                  ClmProcedureCode_3  ClmProcedureCode_4  ClmProcedureCode_5  \
Information Type                                                               
Column Name       ClmProcedureCode_3  ClmProcedureCode_4  ClmProcedureCode_5   
Data Type                    float64             float64             float64   
Sample 1                      4019.0                42.0              4139.0   
Sample 2                      7963.0              4104.0              5185.0   
Sample 3                      2724.0              7999.0              4422.0   
Sample 4                      4439.0              1122.0              27


=== Train_Outpatientdata ===


BeneID    ClaimID  ClaimStartDt  ClaimEndDt  Provider  \
Information Type                                                             
Column Name          BeneID    ClaimID  ClaimStartDt  ClaimEndDt  Provider   
Data Type            object     object        object      object    object   
Sample 1          BENE11002  CLM624349    2009-10-11  2009-10-11  PRV56011   
Sample 2          BENE11003  CLM189947    2009-02-12  2009-02-12  PRV57610   
Sample 3          BENE11004  CLM438021    2009-06-27  2009-06-27  PRV57595   
Sample 4          BENE11005  CLM121801    2009-01-06  2009-01-06  PRV55951   
Sample 5          BENE11006  CLM150998    2009-01-22  2009-01-22  PRV55903   

                  InscClaimAmtReimbursed  AttendingPhysician  \
Information Type                                               
Column Name       InscClaimAmtReimbursed  AttendingPhysician   
Data Type                          int64              object   
Sample 1                              30           PHY326117   
Sample 2                              80           PHY362868   
Sample 3                              10           PHY328821   
Sample 4                              40           PHY334319   
Sample 5                             200           PHY403831   

                  OperatingPhysician  OtherPhysician  ClmDiagnosisCode_1  ...  \
Information Type                                                          ...   
Column Name       OperatingPhysician  OtherPhysician  ClmDiagnosisCode_1  ...   
Data Type                     object          object              object  ...   
Sample 1                   PHY366436       PHY346833               78943  ...   
Sample 2                   PHY359916       PHY311407                6115  ...   
Sample 3                   PHY412248       PHY396473                2723  ...   
Sample 4                   PHY318573       PHY358876               71988  ...   
Sample 5                   PHY319831       PHY394597               82382  ...   

                  ClmDiagnosisCode_9  ClmDiagnosisCode_10  ClmProcedureCode_1  \
Information Type                                                                
Column Name       ClmDiagnosisCode_9  ClmDiagnosisCode_10  ClmProcedureCode_1   
Data Type                     object               object             float64   
Sample 1                       71596                 V059              9672.0   
Sample 2                       34830                V4581              4573.0   
Sample 3                        3310                 2720                66.0   
Sample 4                        4019                  490              5123.0   
Sample 5                        2720                 1970              9390.0   

                  ClmProcedureCode_2  ClmProcedureCode_3  ClmProcedureCode_4  \
Information Type                                                               
Column Name       ClmProcedureCode_2  ClmProcedureCode_3  ClmProcedureCode_4   
Data Type                    float64             float64             float64   
Sample 1                      3811.0               412.0              7840.0   
Sample 2                      7820.0              4401.0               311.0   
Sample 3                      4439.0              4299.0                None   
Sample 4                      4019.0              2724.0                None   
Sample 5                      1741.0                None                None   

                  ClmProcedureCode_5  ClmProcedureCode_6  DeductibleAmtPaid  \
Information Type                                                              
Column Name       ClmProcedureCode_5  ClmProcedureCode_6  DeductibleAmtPaid   
Data Type                    float64             float64              int64   
Sample 1                        None                None                  0   
Sample 2                        None                None                200   
Sample 3                        None                None                 50   
Sample 


=== Train_Beneficiarydata ===


BeneID         DOB         DOD  Gender   Race  \
Information Type                                                     
Column Name          BeneID         DOB         DOD  Gender   Race   
Data Type            object      object      object   int64  int64   
Sample 1          BENE11001  1943-01-01  2009-12-01       1      1   
Sample 2          BENE11002  1936-09-01  2009-09-01       2      2   
Sample 3          BENE11003  1936-08-01  2009-05-01    None      3   
Sample 4          BENE11004  1922-07-01  2009-07-01    None      5   
Sample 5          BENE11005  1935-09-01  2009-08-01    None   None   

                  RenalDiseaseIndicator  State  County  NoOfMonths_PartACov  \
Information Type                                                              
Column Name       RenalDiseaseIndicator  State  County  NoOfMonths_PartACov   
Data Type                        object  int64   int64                int64   
Sample 1                              0     39     230                   12   
Sample 2                              Y     52     280                    0   
Sample 3                           None     24     590                    3   
Sample 4                           None     23     270                    7   
Sample 5                           None     45     680                   10   

                  NoOfMonths_PartBCov  ...  ChronicCond_Depression  \
Information Type                       ...                           
Column Name       NoOfMonths_PartBCov  ...  ChronicCond_Depression   
Data Type                       int64  ...                   int64   
Sample 1                           12  ...                       1   
Sample 2                            0  ...                       2   
Sample 3                            1  ...                    None   
Sample 4                            9  ...                    None   
Sample 5                           11  ...                    None   

                  ChronicCond_Diabetes  ChronicCond_IschemicHeart  \
Information Type                                                    
Column Name       ChronicCond_Diabetes  ChronicCond_IschemicHeart   
Data Type                        int64                      int64   
Sample 1                             1                          1   
Sample 2                             2                          2   
Sample 3                          None                       None   
Sample 4                          None                       None   
Sample 5                          None                       None   

                  ChronicCond_Osteoporasis  ChronicCond_rheumatoidarthritis  \
Information Type                                                              
Column Name       ChronicCond_Osteoporasis  ChronicCond_rheumatoidarthritis   
Data Type                            int64                            int64   
Sample 1                                 2                                1   
Sample 2                                 1                                2   
Sample 3                              None                             None   
Sample 4                              None                             None   
Sample 5                              None                             None   

                  ChronicCond_stroke  IPAnnualReimbursementAmt  \
Information Type                                                 
Column Name       ChronicCond_stroke  IPAnnualReimbursementAmt   
Data Type                      int64                     int64   
Sample 1                           1                     36000   
Sample 2                           2                         0   
Sample 3                        None                      5000   
Sample 4                        None                     21260   
Sample 5                        None                     22000   

                  IPAnnualDeductibleAmt  OPAnnualReimbursementAmt  \
Information Type                                 


=== Train ===


Provider  PotentialFraud
Information Type                          
Column Name       Provider  PotentialFraud
Data Type           object          object
Sample 1          PRV51001              No
Sample 2          PRV51003             Yes
Sample 3          PRV51004            None
Sample 4          PRV51005            None
Sample 5          PRV51007            None

In [3]:
# Inpatient Claims
date_cols_inp = ['ClaimStartDt', 'ClaimEndDt', 'AdmissionDt', 'DischargeDt']
for col in date_cols_inp:
    train_inp[col] = pd.to_datetime(train_inp[col], errors='coerce')

# Outpatient Claims
date_cols_out = ['ClaimStartDt', 'ClaimEndDt']
for col in date_cols_out:
    train_out[col] = pd.to_datetime(train_out[col], errors='coerce')

# Beneficiary Data
date_cols_bene = ['DOB', 'DOD']
for col in date_cols_bene:
    train_bene[col] = pd.to_datetime(train_bene[col], errors='coerce')


In [4]:
print(train_inp[date_cols_inp].dtypes)
print(train_out[date_cols_out].dtypes)
print(train_bene[date_cols_bene].dtypes)


ClaimStartDt    datetime64[ns]
ClaimEndDt      datetime64[ns]
AdmissionDt     datetime64[ns]
DischargeDt     datetime64[ns]
dtype: object
ClaimStartDt    datetime64[ns]
ClaimEndDt      datetime64[ns]
dtype: object
DOB    datetime64[ns]
DOD    datetime64[ns]
dtype: object


In [5]:
train['PotentialFraud'] = train['PotentialFraud'].map({'Yes': 1, 'No': 0})
print(train['PotentialFraud'].value_counts())


PotentialFraud
0    4904
1     506
Name: count, dtype: int64


In [6]:
train_bene['RenalDiseaseIndicator'] = train_bene['RenalDiseaseIndicator'].map({'Y': 1, '0': 0})
print(train_bene['RenalDiseaseIndicator'].value_counts(dropna=False))


RenalDiseaseIndicator
0    118978
1     19578
Name: count, dtype: int64


In [7]:
# Inpatient Claims
train_inp['DeductibleAmtPaid'] = train_inp['DeductibleAmtPaid'].fillna(0)
train_inp['DiagnosisGroupCode'] = train_inp['DiagnosisGroupCode'].fillna('-1')
train_inp['ClmAdmitDiagnosisCode'] = train_inp['ClmAdmitDiagnosisCode'].fillna('Unknown')


In [8]:
# Outpatient Claims
if 'DeductibleAmtPaid' in train_out.columns:
    train_out['DeductibleAmtPaid'] = train_out['DeductibleAmtPaid'].fillna(0)
if 'ClmAdmitDiagnosisCode' in train_out.columns:
    train_out['ClmAdmitDiagnosisCode'] = train_out['ClmAdmitDiagnosisCode'].fillna('Unknown')


In [9]:
print("DeductibleAmtPaid nulls (inpatient):", train_inp['DeductibleAmtPaid'].isnull().sum())
print("DiagnosisGroupCode nulls (inpatient):", train_inp['DiagnosisGroupCode'].isnull().sum())
print("ClmAdmitDiagnosisCode nulls (inpatient):", train_inp['ClmAdmitDiagnosisCode'].isnull().sum())

print("DeductibleAmtPaid nulls (outpatient):", train_out['DeductibleAmtPaid'].isnull().sum())
print("ClmAdmitDiagnosisCode nulls (outpatient):", train_out['ClmAdmitDiagnosisCode'].isnull().sum())


DeductibleAmtPaid nulls (inpatient): 0
DiagnosisGroupCode nulls (inpatient): 0
ClmAdmitDiagnosisCode nulls (inpatient): 0
DeductibleAmtPaid nulls (outpatient): 0
ClmAdmitDiagnosisCode nulls (outpatient): 0


### Data Cleaning Summary

**1. Date Columns**  
All relevant columns in inpatient, outpatient, and beneficiary data converted to pandas datetime.

**2. Target & Indicators**  
- `PotentialFraud` (`Train.csv`): Encoded as 0 (No) and 1 (Yes).
- `RenalDiseaseIndicator` (`Train_Beneficiarydata.csv`): Encoded as 0 (`'0'`) and 1 (`'Y'`).

**3. Numeric & Code Nulls**  
- `DeductibleAmtPaid`: Nulls set to 0 (means no deductible paid).
- `DiagnosisGroupCode`: Nulls set to `-1` (represents “missing/unknown”).
- `ClmAdmitDiagnosisCode`: Nulls set to `'Unknown'`.

**4. High-Missing Columns**  
- All diagnosis, procedure, and physician columns with high missing are retained as-is.
    - Their presence/absence may be informative.
    - Will be addressed further during feature engineering.

**5. All critical columns now have no missing values.**


In [10]:
cleaned_path = '/content/drive/MyDrive/healthcarefraud_detection/data/processed'
train_inp.to_parquet(f'{cleaned_path}/cleaned_inpatient.parquet', index=False)
train_out.to_parquet(f'{cleaned_path}/cleaned_outpatient.parquet', index=False)
train_bene.to_parquet(f'{cleaned_path}/cleaned_beneficiary.parquet', index=False)
train.to_parquet(f'{cleaned_path}/cleaned_train_labels.parquet', index=False)



To ensure the pipeline is production-grade and auditable, we perform the following checks:
- No critical columns have remaining nulls.
- All date logic is valid (e.g., Admission < Discharge).
- All value ranges are valid (e.g., reimbursement amounts ≥ 0).
- Referential integrity is preserved between tables (all keys match).
- Row counts are documented before and after cleaning.
- New flags/features added as required (e.g., HasDied).


In [11]:
# Check critical columns
print(train_inp[['BeneID', 'ClaimID', 'Provider', 'ClaimStartDt', 'ClaimEndDt', 'InscClaimAmtReimbursed']].isnull().sum())
print(train_out[['BeneID', 'ClaimID', 'Provider', 'ClaimStartDt', 'ClaimEndDt', 'InscClaimAmtReimbursed']].isnull().sum())
print(train_bene[['BeneID', 'DOB', 'Gender', 'Race']].isnull().sum())
print(train[['Provider', 'PotentialFraud']].isnull().sum())


BeneID                    0
ClaimID                   0
Provider                  0
ClaimStartDt              0
ClaimEndDt                0
InscClaimAmtReimbursed    0
dtype: int64
BeneID                    0
ClaimID                   0
Provider                  0
ClaimStartDt              0
ClaimEndDt                0
InscClaimAmtReimbursed    0
dtype: int64
BeneID    0
DOB       0
Gender    0
Race      0
dtype: int64
Provider          0
PotentialFraud    0
dtype: int64


In [12]:
# Inpatient: AdmissionDt < DischargeDt
invalid_dates = train_inp[train_inp['AdmissionDt'] > train_inp['DischargeDt']]
print(f"Number of invalid admission/discharge dates: {len(invalid_dates)}")
# Should be zero or very rare—document if any


Number of invalid admission/discharge dates: 0


In [13]:
print("Inpatient negative reimbursement:", (train_inp['InscClaimAmtReimbursed'] < 0).sum())
print("Outpatient negative reimbursement:", (train_out['InscClaimAmtReimbursed'] < 0).sum())
# Should be zero


Inpatient negative reimbursement: 0
Outpatient negative reimbursement: 0


In [14]:
# Providers in claims not present in train labels
inpatient_unmatched = (~train_inp['Provider'].isin(train['Provider'])).sum()
outpatient_unmatched = (~train_out['Provider'].isin(train['Provider'])).sum()
print("Inpatient unmatched providers:", inpatient_unmatched)
print("Outpatient unmatched providers:", outpatient_unmatched)



Inpatient unmatched providers: 0
Outpatient unmatched providers: 0


In [15]:
# BeneIDs in claims not present in master beneficiary table
inpatient_bene_unmatched = (~train_inp['BeneID'].isin(train_bene['BeneID'])).sum()
outpatient_bene_unmatched = (~train_out['BeneID'].isin(train_bene['BeneID'])).sum()
print("Inpatient unmatched BeneID:", inpatient_bene_unmatched)
print("Outpatient unmatched BeneID:", outpatient_bene_unmatched)


Inpatient unmatched BeneID: 0
Outpatient unmatched BeneID: 0


In [16]:
print(f"Train_inp rows: {len(train_inp)}")
print(f"Train_out rows: {len(train_out)}")
print(f"Train_bene rows: {len(train_bene)}")
print(f"Train rows: {len(train)}")
# (Repeat for test data if you clean/test as well)

# If you ever drop rows, print the number dropped and why


Train_inp rows: 40474
Train_out rows: 517737
Train_bene rows: 138556
Train rows: 5410


In [17]:
train_bene['HasDied'] = train_bene['DOD'].notnull().astype(int)
print(train_bene['HasDied'].value_counts())


HasDied
0    137135
1      1421
Name: count, dtype: int64


In [18]:
train_bene.to_parquet('/content/drive/MyDrive/healthcarefraud_detection/data/processed/cleaned_beneficiary.parquet', index=False)


In [19]:
# prompt: Check clean beneficiary data, especially the column, as that and show the top 5 head.

print("\n--- Cleaned Beneficiary Data Head ---")
print(train_bene.head())


--- Cleaned Beneficiary Data Head ---
      BeneID        DOB DOD  Gender  Race  RenalDiseaseIndicator  State  \
0  BENE11001 1943-01-01 NaT       1     1                      0     39   
1  BENE11002 1936-09-01 NaT       2     1                      0     39   
2  BENE11003 1936-08-01 NaT       1     1                      0     52   
3  BENE11004 1922-07-01 NaT       1     1                      0     39   
4  BENE11005 1935-09-01 NaT       1     1                      0     24   

   County  NoOfMonths_PartACov  NoOfMonths_PartBCov  ...  \
0     230                   12                   12  ...   
1     280                   12                   12  ...   
2     590                   12                   12  ...   
3     270                   12                   12  ...   
4     680                   12                   12  ...   

   ChronicCond_Diabetes  ChronicCond_IschemicHeart  ChronicCond_Osteoporasis  \
0                     1                          1                   

In [20]:
from IPython.display import display
import pandas as pd

def transposed_schema_report(df, df_name):
    """
    Generates a transposed tabular report of column names and their types.

    Args:
        df (pd.DataFrame): The DataFrame to report on.
        df_name (str): The name of the DataFrame for the report title.
    """
    print(f"\n--- {df_name} Schema (Transposed) ---")

    # Get the dtypes as a Series
    dtypes_series = df.dtypes

    # Convert the Series to a DataFrame
    # The index (column names) becomes a column, and the values (dtypes) become a column
    schema_df = dtypes_series.reset_index()
    schema_df.columns = ['Column Name', 'Data Type']

    # Set 'Column Name' as the index
    schema_df = schema_df.set_index('Column Name')

    # Transpose the DataFrame
    transposed_df = schema_df.transpose() # The .T accessor also works here

    display(transposed_df)

# Apply the transposed schema report to each DataFrame
transposed_schema_report(train_inp, "Train_Inpatientdata")
transposed_schema_report(train_out, "Train_Outpatientdata")
transposed_schema_report(train_bene, "Train_Beneficiarydata")
transposed_schema_report(train, "Train")


--- Train_Inpatientdata Schema (Transposed) ---


Column Name  BeneID ClaimID    ClaimStartDt      ClaimEndDt Provider  \
Data Type    object  object  datetime64[ns]  datetime64[ns]   object   

Column Name InscClaimAmtReimbursed AttendingPhysician OperatingPhysician  \
Data Type                    int64             object             object   

Column Name OtherPhysician     AdmissionDt  ... ClmDiagnosisCode_7  \
Data Type           object  datetime64[ns]  ...             object   

Column Name ClmDiagnosisCode_8 ClmDiagnosisCode_9 ClmDiagnosisCode_10  \
Data Type               object             object              object   

Column Name ClmProcedureCode_1 ClmProcedureCode_2 ClmProcedureCode_3  \
Data Type              float64            float64            float64   

Column Name ClmProcedureCode_4 ClmProcedureCode_5 ClmProcedureCode_6  
Data Type              float64            float64            float64  

[1 rows x 30 columns]


--- Train_Outpatientdata Schema (Transposed) ---


Column Name  BeneID ClaimID    ClaimStartDt      ClaimEndDt Provider  \
Data Type    object  object  datetime64[ns]  datetime64[ns]   object   

Column Name InscClaimAmtReimbursed AttendingPhysician OperatingPhysician  \
Data Type                    int64             object             object   

Column Name OtherPhysician ClmDiagnosisCode_1  ... ClmDiagnosisCode_9  \
Data Type           object             object  ...             object   

Column Name ClmDiagnosisCode_10 ClmProcedureCode_1 ClmProcedureCode_2  \
Data Type                object            float64            float64   

Column Name ClmProcedureCode_3 ClmProcedureCode_4 ClmProcedureCode_5  \
Data Type              float64            float64            float64   

Column Name ClmProcedureCode_6 DeductibleAmtPaid ClmAdmitDiagnosisCode  
Data Type              float64             int64                object  

[1 rows x 27 columns]


--- Train_Beneficiarydata Schema (Transposed) ---


Column Name  BeneID             DOB             DOD Gender   Race  \
Data Type    object  datetime64[ns]  datetime64[ns]  int64  int64   

Column Name RenalDiseaseIndicator  State County NoOfMonths_PartACov  \
Data Type                   int64  int64  int64               int64   

Column Name NoOfMonths_PartBCov  ... ChronicCond_Diabetes  \
Data Type                 int64  ...                int64   

Column Name ChronicCond_IschemicHeart ChronicCond_Osteoporasis  \
Data Type                       int64                    int64   

Column Name ChronicCond_rheumatoidarthritis ChronicCond_stroke  \
Data Type                             int64              int64   

Column Name IPAnnualReimbursementAmt IPAnnualDeductibleAmt  \
Data Type                      int64                 int64   

Column Name OPAnnualReimbursementAmt OPAnnualDeductibleAmt HasDied  
Data Type                      int64                 int64   int64  

[1 rows x 26 columns]


--- Train Schema (Transposed) ---


Column Name Provider PotentialFraud
Data Type     object          int64

### Production Data Validation

- **Critical columns**: All key fields (IDs, dates, amounts) checked for nulls—none remaining.
- **Date logic**: Validated admission < discharge for inpatient; {N} errors found and documented.
- **Value ranges**: No negative claim amounts detected.
- **Referential integrity**: All providers/BeneIDs in claims exist in master tables; any mismatches are documented.
- **Row counts tracked**: No rows dropped during cleaning.
- **HasDied flag**: Added to beneficiary table (1 if DOD present, else 0).
- **Schema validation**: Final dtypes printed and confirmed correct.



### Key Findings Summary

Here is a concise summary of the important findings and statistics from the data loading and cleaning process:

*   **Potential Fraud Distribution (Train Data):**
    *   No: 4904
    *   Yes: 506
*   **Renal Disease Indicator Distribution (Train Beneficiary Data):**
    *   0: 118978
    *   1: 19578
*   **Missing Value Imputation (Train Data):** All critical columns and specifically imputed columns (`DeductibleAmtPaid`, `DiagnosisGroupCode`, `ClmAdmitDiagnosisCode`) show 0 nulls remaining after cleaning.
*   **Invalid Admission/Discharge Dates (Train Inpatient Data):** 0 records found with invalid date ranges.
*   **Negative Reimbursement Amounts (Train Data):** 0 records found with negative claim reimbursement amounts.
*   **Referential Integrity Checks (Train Data):** All Provider and Beneficiary IDs in claim data successfully matched to their respective master tables.
*   **Row Counts (Train Data):** Row counts remain unchanged, indicating no rows were dropped during this cleaning phase:
    *   `train_inp`: 40474 rows
    *   `train_out`: 517737 rows
    *   `train_bene`: 138556 rows
    *   `train`: 5410 rows
*   **Has Died Flag Distribution (Train Beneficiary Data):**
    *   0 (Did not die): 137135
    *   1 (Died): 1421


##streamlit

In [21]:
# Example for inpatient data
train_inp.head(0).T.to_csv('/content/drive/MyDrive/healthcarefraud_detection/streamlit_app/data/train_inp_schema.csv')
train_out.head(0).T.to_csv('/content/drive/MyDrive/healthcarefraud_detection/streamlit_app/data/train_out_schema.csv')
train_bene.head(0).T.to_csv('/content/drive/MyDrive/healthcarefraud_detection/streamlit_app/data/train_bene_schema.csv')
train.head(0).T.to_csv('/content/drive/MyDrive/healthcarefraud_detection/streamlit_app/data/train_schema.csv')


In [22]:
cleaning_summary = """
Data Cleaning Summary

1. Date Columns
All relevant columns in inpatient, outpatient, and beneficiary data converted to pandas datetime.

2. Target & Indicators
- PotentialFraud (Train.csv): Encoded as 0 (No) and 1 (Yes).
- RenalDiseaseIndicator (Train_Beneficiarydata.csv): Encoded as 0 ('0') and 1 ('Y').

3. Numeric & Code Nulls
- DeductibleAmtPaid: Nulls set to 0 (means no deductible paid).
- DiagnosisGroupCode: Nulls set to -1 (represents “missing/unknown”).
- ClmAdmitDiagnosisCode: Nulls set to 'Unknown'.

4. High-Missing Columns
- All diagnosis, procedure, and physician columns with high missing are retained as-is.
- Their presence/absence may be informative.
- Will be addressed further during feature engineering.

Checks for:
- No critical columns have remaining nulls.
- All date logic is valid (e.g., Admission < Discharge).
- All value ranges are valid (e.g., reimbursement amounts ≥ 0).
- Referential integrity is preserved between tables (all keys match).
- Row counts are documented before and after cleaning.
- New flags/features added as required (e.g., HasDied).
"""
with open("/content/drive/MyDrive/healthcarefraud_detection/streamlit_app/data/cleaning_summary.txt", "w") as f:
    f.write(cleaning_summary)
print("Cleaning summary saved!")


Cleaning summary saved!


In [23]:
schema_info = {col: str(dtype) for col, dtype in zip(train_inp.columns, train_inp.dtypes)}
import json
with open("/content/drive/MyDrive/healthcarefraud_detection/streamlit_app/data/train_inp_schema.json", "w") as f:
    json.dump(schema_info, f, indent=4)
print("train_inp_schema.json saved!")


train_inp_schema.json saved!


In [24]:
checks = {
    "train_inp rows": len(train_inp),
    "train_out rows": len(train_out),
    "train_bene rows": len(train_bene),
    "train rows": len(train),
    "inpatient unmatched providers": int((~train_inp['Provider'].isin(train['Provider'])).sum()),
    "outpatient unmatched providers": int((~train_out['Provider'].isin(train['Provider'])).sum()),
    "inpatient unmatched BeneID": int((~train_inp['BeneID'].isin(train_bene['BeneID'])).sum()),
    "outpatient unmatched BeneID": int((~train_out['BeneID'].isin(train_bene['BeneID'])).sum())
}
with open("/content/drive/MyDrive/healthcarefraud_detection/streamlit_app/data/cleaning_checks.json", "w") as f:
    json.dump(checks, f, indent=4)
print("Sanity check file saved!")


Sanity check file saved!


In [25]:
train_inp.head(10).to_csv('/content/drive/MyDrive/healthcarefraud_detection/streamlit_app/data/train_inp_cleaned_sample.csv', index=False)
train_out.head(10).to_csv('/content/drive/MyDrive/healthcarefraud_detection/streamlit_app/data/train_out_cleaned_sample.csv', index=False)
train_bene.head(10).to_csv('/content/drive/MyDrive/healthcarefraud_detection/streamlit_app/data/train_bene_cleaned_sample.csv', index=False)
